Service authorization

noseglid opened the pull request:
This change adds support for [Fastly Service Authorizations][1] to be
managed via Terraform.

Other managed users and services can be referenced by their ID and the
permission is controlled with an enumeration of the allowed values by
the API (`full`, `read_only`, `purge_select` or `purge_all`).

[1]: https://developer.fastly.com/reference/api/account/service-authorization/

noseglid committed:
Add support for Service Authorizations

This change adds support for [Fastly Service Authorizations][1] to be
managed via Terraform.

Other managed users and services can be referenced by their ID and the
permission is controlled with an enumeration of the allowed values by
the API (`full`, `read_only`, `purge_select` or `purge_all`).

[1]: https://developer.fastly.com/reference/api/account/service-authorization/

noseglid committed:
Service Authorizations: Adds docs

Add documentation on how to use and import services authorizations with
Fastly.

noseglid commented on go.mod:
```diff
@@ -12,3 +12,5 @@ require (
 	golang.org/x/net v0.0.0-20211015210444-4f30a5c0130f
 )
+
+replace github.com/fastly/go-fastly/v6 => github.com/noseglid/go-fastly/v6 v6.0.0-20220615131558-6b72cc694806
```
This needs to be replaced before this PR is merged. It requires https://github.com/fastly/go-fastly/pull/349 to be merged and released first.

Integralist commented on examples/resources/service_authorization_basic_usage.tf:
```diff
@@ -0,0 +1,11 @@
+resource "fastly_service_vcl" "myservice" {
```
Can we add some example fields to each section, as this will make it more consistent with the other example files. For example, the [ACL Entries example](https://github.com/fastly/terraform-provider-fastly/blob/78fe6303e670d18a21884cabc610a93a9272a748/examples/resources/service_acl_entries_basic_usage.tf) provides configuration for all resources defined to make the example as complete as possible.

Integralist replied:
I would also rename `"myservice"` to `"demo"` so the following `examples/resources/service_authorization_import.txt` file will align.

> **NOTE**: Our use of `myservice` and `demo` are a bit inconsistent, but looks like `demo` is used more.

noseglid replied:
Sure, I will change to `demo`.

Integralist commented on fastly/resource_fastly_service_authorization.go:
```diff
@@ -0,0 +1,115 @@
+			"permission": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				Default:          "full",
```
I'm actually surprised the API _defaults_ to "full". 

I feel it would be safer to default to "read_only" and have users explicitly opt into more permissive values. 

> **NOTE**: I've reached out to the relevant API team about this as I'm interested to know if they agree that we should default to "read_only" (and if so, then we should update the API endpoint as well to reflect that).

noseglid replied:
I agree. `read_only` would absolutely be a better default.

Alternatively, there could be *no default*, and explicitly require users to set the permissions they desire. Better to be explicit in my opinion.

Integralist replied:
Agreed. Let's avoid the default and users have to be explicit about what they want. 👍🏻

Integralist commented on fastly/resource_fastly_service_authorization.go:
```diff
@@ -0,0 +1,115 @@
+func resourceServiceAuthorizationUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*FastlyClient).conn
+
+	// Update Name and/or Role.
```
Not sure what this line refers to? 

I think it's a side-effect of copy/pasting from another file (`fastly/resource_fastly_user.go`?)

noseglid replied:
That's exactly it 😅  I'll sort it.

Integralist reviewed (commented):
Thanks @noseglid for opening this PR.

Other than a couple of minor comments I have (and also the changes you mentioned regarding ensuring we use a new go-fastly release, once your associated PR is merged/released) this PR LGTM.

noseglid committed:
service_authorization: Align demo resource names

noseglid committed:
service_authorization: Require permission

The default permission in the API is `full` which is extremely
permissive and arguably not a good default. The terraform provider will
require the permission to be set to avoid over-permitting users by a
simple omission.

noseglid committed:
go.mod: Bump version of go-fastly

noseglid committed:
service_authorization: Add acceptance test

Acceptance tests ensure the resource works as expected right now, but
also that the functionality keeps working as other changes are applied.

noseglid committed:
service_authorization: Re-generate docs

noseglid committed:
Use main branch of fastly/go-fastly

Removes the temporary replace directive. This should still be replaced
with a tag once the changes on main have actually been released.

noseglid commented:
This PR is now updated using the main branch of go-fastly/fastly. I will update it again once we have a tagged version to use.

Integralist commented:
@noseglid I've cut a new go-fastly release now https://github.com/fastly/go-fastly/releases/tag/v6.4.0

noseglid committed:
go.mod: Use released version of github.com/fastly/go-fastly/v6

noseglid commented:
@Integralist Any chance to get a new release with these changes included?

Integralist commented:
Hi @noseglid 

Absolutely. I was planning on cutting a release after merging this PR but it was then brought to my attention (via another customer) that the go-fastly change set was missing the list endpoint and that took up my focus for a bit.

So (in case it's of interest to you) I've since added support for listing service authorizations, and also cut that into a new go-fastly release:
https://github.com/fastly/go-fastly/releases/tag/v6.4.1

I can't cut a new Terraform release unless all integration tests are passing, which at the moment there are a couple of failing tests related to changes made in our API layer, so I have a PR open that fixes those, and I'm just waiting a member of my team to review/approve:
https://github.com/fastly/terraform-provider-fastly/pull/573

Once that's done I'll cut a new `v2.1.0` release and ping here to let you know when it's done.

noseglid commented:
That sounds great. Thank you for the update!

Integralist commented:
@noseglid Sorry for the delay:
https://github.com/fastly/terraform-provider-fastly/releases/tag/v2.1.0

noseglid commented:
Fantastic, thank you!
